File: course_4/module_4/examples/two_sat.py
import random
from typing import List, Tuple


class TwoSAT:
    def __init__(self, num_variables: int, clauses: List[Tuple[int, int]], seed: int = None, debug: bool = False):
        """
        Initializes a 2-SAT problem instance.
        :param num_variables: Number of boolean variables in the problem.
        :param clauses: List of clauses, where each clause is a tuple of two literals.
                        Positive integers represent variables (e.g., 1 means x1 is True),
                        Negative integers represent negated variables (e.g., -1 means x1 is False).
        :param seed: Seed for random number generation. Used for reproducibility.
        :param debug: If True, prints debug information during execution.
        """
        self.num_variables = num_variables
        self.clauses = clauses
        self.debug = debug
        if seed is not None:
            random.seed(seed)  # Set the random seed if provided for reproducible results.

        # Validate that all literals in clauses reference valid variable indices.
        if any(abs(literal) > num_variables for clause in clauses for literal in clause):
            raise ValueError("Clause contains variable out of bounds.")

    def is_satisfiable(self) -> Tuple[bool, List[bool]]:
        """
        Determines whether the 2-SAT problem is satisfiable using randomized local search.
        :return: A tuple (is_satisfiable, assignment).
                 is_satisfiable: True if satisfiable, False otherwise.
                 assignment: A list representing the variable assignments if satisfiable, else empty.
        """
        # Set the maximum number of flips and trials for the algorithm.
        max_attempts = 2 * self.num_variables ** 2  # Maximum flips per trial.
        max_trials = 10 * self.num_variables  # Maximum number of independent trials.

        # Perform up to max_trials independent runs of the local search algorithm.
        for trial in range(max_trials):
            # Generate a random initial assignment for all variables.
            assignment = [random.choice([True, False]) for _ in range(self.num_variables)]

            # Perform up to max_attempts flips to try to satisfy all clauses.
            for flip in range(max_attempts):
                # Identify all clauses that are not currently satisfied.
                unsatisfied = [clause for clause in self.clauses if not self._is_clause_satisfied(clause, assignment)]

                if not unsatisfied:  # If no unsatisfied clauses remain, the problem is satisfiable.
                    return True, assignment

                # Randomly choose an unsatisfied clause.
                clause = random.choice(unsatisfied)

                # Randomly choose one of the two literals in the clause and flip its value.
                variable_to_flip = abs(random.choice(clause)) - 1  # Convert to 0-based index.
                assignment[variable_to_flip] = not assignment[variable_to_flip]

                # If debug mode is enabled, print the current trial, flip, and assignment.
                if self.debug:
                    print(f"Trial {trial}, Flip {flip}, Assignment: {assignment}")

        # If no satisfying assignment is found after all trials, the problem is unsatisfiable.
        return False, []

    @staticmethod
    def _is_clause_satisfied(clause: Tuple[int, int], assignment: List[bool]) -> bool:
        """
        Checks whether a given clause is satisfied by the current assignment.
        :param clause: A tuple of two literals (e.g., (1, -2)).
        :param assignment: A list of current assignments for all variables.
        :return: True if the clause is satisfied, False otherwise.
        """
        u, v = clause  # Extract the two literals from the clause.
        # Determine the value of the first literal based on its sign.
        value_u = assignment[abs(u) - 1] if u > 0 else not assignment[abs(u) - 1]
        # Determine the value of the second literal based on its sign.
        value_v = assignment[abs(v) - 1] if v > 0 else not assignment[abs(v) - 1]
        # The clause is satisfied if either literal is True.
        return value_u or value_v


# Example usage:
if __name__ == "__main__":
    # Example 2-SAT problem with 4 variables and 4 clauses.
    num_vars = 4  # Number of variables.
    example_clauses = [
        (1, 2),     # x1 OR x2
        (-1, 3),    # NOT x1 OR x3
        (3, 4),     # x3 OR x4
        (-2, -4)    # NOT x2 OR NOT x4
    ]

    # Create a TwoSAT solver instance with debug mode enabled.
    two_sat_solver = TwoSAT(num_vars, example_clauses, debug=True)
    # Run the solver to check satisfiability.
    is_satisfiable, assignment = two_sat_solver.is_satisfiable()

    if is_satisfiable:
        # If satisfiable, print the satisfying assignment for each variable.
        print(f"The 2-SAT problem is satisfiable with the assignment:")
        for i, value in enumerate(assignment, start=1):
            print(f"x{i} = {'True' if value else 'False'}")
    else:
        # If unsatisfiable, print a corresponding message.
        print("The 2-SAT problem is unsatisfiable.")

File: course_4/module_4/examples/two_sat.go
package main

import (
	"fmt"
	"math/rand"
)

// TwoSAT represents a 2-SAT problem instance.
type TwoSAT struct {
	numVariables int      // Number of variables in the problem.
	clauses      [][2]int // List of clauses, where each clause is a pair of literals.
	debug        bool     // If true, enables debug output.
}

// NewTwoSAT initializes a new 2-SAT problem instance.
func NewTwoSAT(numVariables int, clauses [][2]int, debug bool) (*TwoSAT, error) {
	// Validate that all literals in clauses are within the valid range of variables.
	for _, clause := range clauses {
		for _, literal := range clause {
			if abs(literal) > numVariables {
				return nil, fmt.Errorf("invalid literal %d: exceeds number of variables", literal)
			}
		}
	}
	return &TwoSAT{
		numVariables: numVariables,
		clauses:      clauses,
		debug:        debug,
	}, nil
}

// IsSatisfiable determines whether the 2-SAT problem instance is satisfiable.
// It uses randomized local search to find a satisfying assignment if it exists.
func (ts *TwoSAT) IsSatisfiable() (bool, []bool) {
	maxAttempts := 2 * ts.numVariables * ts.numVariables // Maximum flips per trial.
	maxTrials := 10 * ts.numVariables                    // Maximum number of independent trials.

	// Perform up to maxTrials independent runs of the local search algorithm.
	for trial := 0; trial < maxTrials; trial++ {
		// Generate a random initial assignment for all variables.
		assignment := make([]bool, ts.numVariables)
		for i := range assignment {
			assignment[i] = rand.Intn(2) == 1 // Randomly set each variable to true or false.
		}

		// Perform up to maxAttempts flips to try to satisfy all clauses.
		for flip := 0; flip < maxAttempts; flip++ {
			// Identify unsatisfied clauses.
			unsatisfied := ts.getUnsatisfiedClauses(assignment)
			if len(unsatisfied) == 0 {
				// If no unsatisfied clauses remain, the problem is satisfiable.
				return true, assignment
			}

			// Randomly choose an unsatisfied clause.
			clause := unsatisfied[rand.Intn(len(unsatisfied))]

			// Randomly choose one of the two literals in the clause and flip its value.
			variableToFlip := abs(clause[rand.Intn(2)]) - 1
			assignment[variableToFlip] = !assignment[variableToFlip]

			// Print debug information if enabled.
			if ts.debug {
				fmt.Printf("Trial %d, Flip %d, Assignment: %v\n", trial, flip, assignment)
			}
		}
	}

	// If no satisfying assignment is found after all trials, the problem is unsatisfiable.
	return false, nil
}

// getUnsatisfiedClauses returns a list of unsatisfied clauses for the current assignment.
func (ts *TwoSAT) getUnsatisfiedClauses(assignment []bool) [][2]int {
	unsatisfied := [][2]int{}
	for _, clause := range ts.clauses {
		if !ts.isClauseSatisfied(clause, assignment) {
			unsatisfied = append(unsatisfied, clause)
		}
	}
	return unsatisfied
}

// isClauseSatisfied checks whether a given clause is satisfied by the current assignment.
func (ts *TwoSAT) isClauseSatisfied(clause [2]int, assignment []bool) bool {
	u, v := clause[0], clause[1]
	valueU := assignment[abs(u)-1]
	if u < 0 {
		valueU = !valueU
	}
	valueV := assignment[abs(v)-1]
	if v < 0 {
		valueV = !valueV
	}
	return valueU || valueV
}

// Helper function to compute the absolute value of an integer.
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	// Example 2-SAT problem with 4 variables and 4 clauses.
	numVars := 4
	clauses := [][2]int{
		{1, 2},   // x1 OR x2
		{-1, 3},  // NOT x1 OR x3
		{3, 4},   // x3 OR x4
		{-2, -4}, // NOT x2 OR NOT x4
	}

	// Initialize the TwoSAT solver.
	twoSAT, err := NewTwoSAT(numVars, clauses, true)
	if err != nil {
		fmt.Printf("Error initializing 2-SAT: %v\n", err)
		return
	}

	// Check satisfiability.
	isSatisfiable, assignment := twoSAT.IsSatisfiable()
	if isSatisfiable {
		// Print the satisfying assignment if one is found.
		fmt.Println("The 2-SAT problem is satisfiable with the following assignment:")
		for i, value := range assignment {
			fmt.Printf("x%d = %v\n", i+1, value)
		}
	} else {
		// Print a message indicating that the problem is unsatisfiable.
		fmt.Println("The 2-SAT problem is unsatisfiable.")
	}
}
